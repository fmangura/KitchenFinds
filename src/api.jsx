import React, {useState} from 'react'
import axios from 'axios'
import qs from 'qs';


import Cookies from 'universal-cookie';
const cookies = new Cookies(null, { path: '/' });

const BACKEND_API_URI = 'http://localhost:3001';

class backend {  
    static user_token = cookies.get('user_token') || ""

    // General request to automatically include token 

    static async request(endpoint, data = {}, method = "GET") {
        const url = `${BACKEND_API_URI}${endpoint}`;
        const headers = { authorization: `Bearer ${backend.user_token}`};
        const params = (method === "GET")
            ? data
            : {};

        try {
            return (await axios({ url, method, data, params, headers })).data;
        } catch (err) {
            console.error("Backend API Error:", err.response);
        }
    }

    // Users individual routes

    static async register(data) {
        return await backend.request(`/users/register`, data, 'post').then((res => {message: 'Register Complete!'}))
    }

    static async login(formData) {
        try {
            return await this.request(`/users/login`, formData, 'post')
                .then((res) => {
                    let {token, user} = res
                    cookies.set('user_token', token)
                    cookies.set('user', user)
                    backend.user_token = cookies.get('user_token') || ""
                    return user
                })
        } catch (err) {
            console.log(err.data)
            return err.data
        }
    }

    static async delUser(formData) {
        try {
            return await this.request(`/users/${cookies.get('user')}`, formData, 'delete')
                .then((res) => res)
        } catch (err) {
            return err
        }
    }

    static async getUserRecipes() {
        try{
            return await backend.request(`/users/myrecipes`).then((res) => res)
        } catch (err) {
            return err
        }
    }

    // Recipes individual routes

    static async getRecipeIngredients(ref_id) {
        return await backend.request(`/recipes/getRecipe/${ref_id}`)
    }

    static async searchByIngredients(items) {
        try {
            const ingrList = qs.stringify({ ingredients: items }, { arrayFormat: 'comma' })
            return await axios({
                url: `${BACKEND_API_URI}/recipes/search?${ingrList}`,
                method: "GET",
                headers: { authorization: `Bearer ${backend.user_token}`},
            }).then((res) => res)
        } catch (err) {
            return err
        }
    }

    static async getInstructions(ref_id) {
        try {
            return await this.request(`/recipes/instructions/${ref_id}`).then((res) => {console.log(res); return res})
        } catch (err) {
            return err
        }
        
    }

    static async addToFave(ref_id) {
        try {
            return await backend.request(`/recipes/${ref_id}/fave`, {},'post').then((res) => res)
        } catch (err) {
            return err
        }
    }

    static async delFave(ref_id) {
        try {
            return await backend.request(`/recipes/${ref_id}`, {}, 'delete').then((res) => res)
        } catch (err) {
            return err
        }
    }

    static async getSuggestions(search) {
        try {
            return await axios({
                url:`${BACKEND_API_URI}/recipes/suggestions`, 
                data: {'item': search},
                method:'POST'
            }).then((res) => {console.log(res.data); return res.data})
        } catch (err) {
            return err
        }
    }
}

export default backend;

