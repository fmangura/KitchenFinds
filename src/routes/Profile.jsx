import React, {useContext, useEffect, useState} from 'react';
import { Navigate, useNavigate } from 'react-router-dom';
import { AuthContext } from '../context';
import backend from '../api'
import IngredientsTag from '../components/IngredientTag';
import useForm from '../helper/userForm';

import Rat from '../assets/rat.svg'
import Foods from '../assets/8.svg'
import Vegs from '../assets/9.svg'
import Hangfruit from '../assets/hangfruit.svg'
import Celery from '../assets/celery.svg'

import './profile.css'

export default function Profile () {
    const navigate = useNavigate();
    const {isLoggedIn, logout} = useContext(AuthContext)
    const [recipes, setRecipes] = useState([])
    const [hidden, setHidden] = useState(true)
    const initialForm = {
        password: ''
    }
    const {formData, handleChange, resetForm} = useForm(initialForm)
    
    useEffect(() => {
        const addRecipes = async () => {
            let res = await backend.getUserRecipes()
            if (res) setRecipes(recipes => res)
        }
        addRecipes()
    },[])
    const [errMsg, setErrMsg] = useState('')

    const deleteUser = async (e) => {
        e.preventDefault()
        try {
            await backend.delUser(formData).then()
            logout()
        } catch (err) {
            return err
        }
    }
    const delCancel = () => {
        resetForm()
        setHidden(true)
    }

    if (!isLoggedIn) return <Navigate to={'/'}/>

    return (
        <div className='userRecipes'>
            <img className='background-images' id='foods' src={Foods} />
            <img className='background-images' id='vegs' src={Vegs} />
            <img className='background-images' id='hangfruit' src={Hangfruit} />
            <img className='background-images' id='celery' src={Celery} />
            {recipes[0] != null ?
                <div className='user-recipe-list'>
                {recipes.map(({id, name, allIngr, img}, idx) =>  
                    <div className='card userRecipe-card'>
                        <div className='information'>
                            <h1>{name}</h1>
                            <ul className='user-ingr'>
                                {allIngr.map(ingr => {return <li>{ingr}</li>})}
                            </ul>
                            <small>{idx+1}</small>
                        </div>
                        <div className='image'>
                            {img ? <img src={img} alt='No image found' /> : <><img src={Rat} /><p>No Image Found</p></>}
                        </div>
                    </div>
                )} 
                </div> 
                :
                <div className='user-recipe-list' style={{width:'60%'}}>
                    <h1>No saved recipes found.</h1>
                </div>
            }
            <a onClick={() => {hidden ? setHidden(false) : setHidden(true)}} className='delete-button'>Delete Profile</a>
            <div style={hidden ? {display:'none'} : {}} className='del-overlay'>
                <div className='card-del-profile card card--accent'>
                    <h3>Want to delete your pofile?</h3>
                    <p>Please input your password to complete request.</p>
                    <form>
                    <label htmlFor='password'>Password: </label>
                    <input type='password' name='password' placeholder='password' id='password' onChange={handleChange} value={formData.password}></input>
                    <div style={{display:'flex', marginTop:'30px', justifyContent:'space-between'}}>
                        <button onClick={delCancel}>Cancel</button>
                        <a onClick={deleteUser} id='del-complete'>Delete</a>
                    </div>
                    </form>
                </div>
            </div>
        </div>
    );
}