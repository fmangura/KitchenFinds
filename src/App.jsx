import React, {useState} from 'react'
import { BrowserRouter, Routes, Route } from 'react-router-dom';

import Auth from './components/AuthProvider'
import Home from './routes/Home'
import LoginForm from './routes/LoginForm'
import RegisterForm from './routes/RegisterForm'
import Nav from './components/Nav';
import Profile from './routes/Profile'

export default function App () {

    return (
        <>
            <BrowserRouter>
                <Auth>
                    <Nav />
                    <Routes>
                        <Route path='/' element={<Home />} />
                        <Route path='/profile' element={<Profile />} />
                        <Route path='/login' element={<LoginForm />} />
                        <Route path='/register' element={<RegisterForm />} />
                    </Routes>
                </Auth>
            </BrowserRouter>
        </>
    );
}