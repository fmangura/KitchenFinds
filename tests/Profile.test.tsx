import React from 'react';
import { AuthContext } from '../src/context';
import {fireEvent, getByText, queryAllByText, render, screen} from '@testing-library/react'
import '@testing-library/jest-dom'
import Profile from '../src/routes/Profile'
import backend from '../src/api';
import { vi } from 'vitest';
import { Router, Routes, BrowserRouter } from 'react-router-dom';

const mockedUseNavigate = vi.fn();
vi.mock("react-router-dom", async () => {
  const mod = await vi.importActual<typeof import("react-router-dom")>(
    "react-router-dom"
  );
  return {
    ...mod,
    useNavigate: () => mockedUseNavigate,
  };
});

const renderComponent = () => {
    const value = {currUser:'u1', isLoggedIn:'true'}
    return render(
        <AuthContext.Provider value={value} >
                <Profile />
        </AuthContext.Provider>
    );
}

describe('Profile route works', () => {
    it('renders without crashing with Name', () => {
        renderComponent();
        const button = screen.queryByText('Delete Profile')
        expect(button).toBeInTheDocument()
    });

    it('checks snapshot', () => {
        const {asFragment} = renderComponent();
        expect(asFragment).toMatchSnapshot();
    })

    it('Can see delete form', () => {
        renderComponent();
        const delbutton = screen.getByText('Delete Profile')
        fireEvent.click(delbutton)
        expect(screen.getByText('Please input your password to complete request.')).toBeVisible()
    });

})