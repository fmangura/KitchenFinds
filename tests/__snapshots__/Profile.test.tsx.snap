// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Profile route works > checks snapshot 1`] = `[Function]`;
